### DISCUSSION TITLE
Replace QuillJs in the RichTextEdit?

### DISCUSSION THREAD

**Discussion by njannink** [Megabit/Blazorise#1908] [General]
I was doing some investigation and the last release that we use from QuillJs is from sep 2019. Apparantly the maintainer is now focusing on his commercial product see this [discussion](https://github.com/quilljs/quill/issues/3251).

We might consider swapping the editor from something that is more actively maintained eg

- [TinyMCE](https://www.tiny.cloud/features)
- [ProseMirror](https://prosemirror.net/)
- [Trix](https://trix-editor.org/)
- [SummerNote](https://summernote.org/)
- [DevExtreme Quill (fork)](https://github.com/DevExpress/devextreme-quill)

Im thinking about TinyMCE because that one is actively being maintained and used

**Comment by njannink** [comment]
Maybe a new extension `Blazorise.RichTextEdit.TinyMCE` needs to be created to not immediately break the old API

**Reply by stsrki** [reply]
If it's not maintained anymore I don't see a reason to further continue with supporting it. We can leave it as it is for now but in the long run, we can create `Blazorise.RichTextEdit` base project and then have implementation for each supported provider(TinyMCE, ProseMirror, or even QuillJs). Just like we have for Bootstrap, Bulma, etc.

But it's a breaking change so we will need to plan it well.

**Reply by njannink** [reply]
Indeed thats why I suggested to make it into a new extension and keep the QuillJS based variant in the current extension and possibly in a release later rename that extension to `Blazorise.RichTextEdit.QuillJS`

**Reply by njannink** [reply]
I will work that out. With the dynamic loading concept that I use for QuillJS it should work fine

**Comment by ghost** [comment]
Note that TinyMCE has more restrictive licencing than QuillJS.

I wasn't able to use Tiny because I am creating an editor which should be able to work offline, and with Tiny I only get 30 days development time free before I have to licence in that scenario, whereas Quill doesn't have that restriction.

**Reply by stsrki** [reply]
Yeah, I'm aware. But that all depends on the user of the library. We can just give you a tool to work with Blazor implementation. You are responsible for appropriate licensing.

**Reply by njannink** [reply]
The community version of TinyMCE seems fine see: https://www.tiny.cloud/pricing/

@Melestrua what are you using then for editor? QuillJS?

**Reply by stsrki** [reply]
It might be a good idea to have a _licensing mechanism_ for libraries that have them.

**Reply by njannink** [reply]
I don't think we will need it ourselfs. TinyMCE has it's own licensing/deployment mechanism for it's premium features. Only thing is we should not add support for turning off the branding, because in the community edition you are not allowed to turn that off.

**Reply by stsrki** [reply]
No, I didn't mean to turn off anything. I just want to make licensing easier. For example, instead of adding an API key of any libraries into the index.html file, we could have this:

```cs
services.AddRichTextEdit(options=>{
  options.LicenceKey = "abc";
});
```

and then it would all be done automatically by the Blazorise extension.

**Reply by ghost** [reply]
> The community version of TinyMCE seems fine see: https://www.tiny.cloud/pricing/

It requires you to use the cloud version of TinyMCE, which doesn't work for me because I want to be able to run completely offline (I had the discussion with their licencing team). Yes, I've been using QuillJS.

**Reply by ghost** [reply]
TBH my use case is probably a bit out of band. I'm experimenting creating a multi-page cross-platform app with rich text edit including understanding/translating typical markdown commands, adding cross-references inline from a menu and save to/restore from markdown.

With all the additions/plugins, QuillJS seems to offer the features. But I have to confess I'm wondering whether Blazor is the best fit since web isn't my primary target, it's just where the rich text/markdown seems to be best supported.

**Comment by swaipa** [comment]
You could take a look at Trix by Basecamp (Ruby on Rails creators).

https://trix-editor.org/
https://github.com/basecamp/trix

**Reply by stsrki** [reply]
Trix editor looks really nice.

**Reply by njannink** [reply]
I find it a bit limited. No support for emoijs for example
